### PR TITLE
Add tests to Verify CPU and Memory requests for the CUDA and Image Pods(ODS-313)

### PR DESCRIPTION
**Pull request by prudhvi-165**
Signed-off-by: prudhvi <pnammi@redhat.com>

### CODE REVIEW

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/1001/
Result: Keyword Added works fine

File: tests/Tests/100__deploy/100__installation/102__post_install.robot [hunk 1]
```diff
@@ -315,3 +317,20 @@ Replace "Prometheus" With "Grafana" In Rhods-Monitor-Federation
     OpenShiftLibrary.Oc Patch    kind=ServiceMonitor
     ...                   src={"spec":{"selector":{"matchLabels": {"app":"grafana"}}}}
     ...                   name=rhods-monitor-federation   namespace=redhat-ods-monitoring  type=merge
+
+Verify CPU And Memory Requests For Image And Base Cuda Pods
+    [Documentation]     Verifies CPU and Memory Requests for image and base cuda pods
+    [Arguments]   ${namespace}=redhat-ods-applications
+    ${pods} =  Oc Get  kind=Pod  namespace=${namespace}
+    FOR    ${pod}    IN  @{pods}
```
Comment (bdattoma) [inline tests/Tests/100__deploy/100__installation/102__post_install.robot:325]: could you try fetching only the pods of cuda using the filters in Oc get?

File: tests/Tests/100__deploy/100__installation/102__post_install.robot [hunk 2]
```diff
@@ -315,3 +317,20 @@ Replace "Prometheus" With "Grafana" In Rhods-Monitor-Federation
     OpenShiftLibrary.Oc Patch    kind=ServiceMonitor
     ...                   src={"spec":{"selector":{"matchLabels": {"app":"grafana"}}}}
     ...                   name=rhods-monitor-federation   namespace=redhat-ods-monitoring  type=merge
+
+Verify CPU And Memory Requests For Image And Base Cuda Pods
+    [Documentation]     Verifies CPU and Memory Requests for image and base cuda pods
```
Comment (bdattoma) [inline tests/Tests/100__deploy/100__installation/102__post_install.robot:322]: not clear what "image and base cuda" refers to?

Comment (prudhvi-165) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:322]: Image -> pytorch,tensorflow,minimal-gpu
Cuda-> 11.4.2-cuda-s2i-base-ubi8-1-build,11.4.2-cuda-s2i-core-ubi8-1-build,11.4.2-cuda-s2i-py38-ubi8-1-build,11.4.2-cuda-s2i-thoth-ubi8-py38-1-build

File: tests/Tests/100__deploy/100__installation/102__post_install.robot (hunk 1 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:325]: But we still need to fetch image pods as well ,separately. Why not fetch all of them at once and filter one after the other?

Comment (bdattoma) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:325]: ok

File: tests/Tests/100__deploy/100__installation/102__post_install.robot (hunk 2 above)
Comment (bdattoma) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:322]: `Verify CPU And Memory Requests In Pods For Builds` is a suggestion for a more clear name, maybe

File: tests/Tests/100__deploy/100__installation/102__post_install.robot [hunk 3]
```diff
@@ -315,3 +317,20 @@ Replace "Prometheus" With "Grafana" In Rhods-Monitor-Federation
...
+        ...    Verify Requests Contains Expected Values  cpu=2  memory=4Gi  requests=${pod['spec']['containers'][0]['resources']['requests']}
+        Run Keyword If    "minimal-gpu" in "${pod['metadata']['name']}" or "pytorch" in "${pod['metadata']['name']}" or "tensorflow" in "${pod['metadata']['name']}"
+        ...    Verify Requests Contains Expected Values  cpu=4  memory=8Gi  requests=${pod['spec']['containers'][0]['resources']['requests']}
+    END
+
+Verify Requests Contains Expected Values
+    [Documentation]     Verifies cpu and memory requests contain expected values
+    [Arguments]   ${cpu}  ${memory}  ${requests}
+    Should Be Equal As Strings    ${requests['cpu']}  ${cpu}
+    Should Be Equal As Strings    ${requests['memory']}  ${memory}
```
Comment (lugi0) [inline tests/Tests/100__deploy/100__installation/102__post_install.robot:336]: Couldn't you reuse the `Verify CPU And Memory Requests And Limits Are Defined For All Containers In All Pods in Project` keyword here?
it loops over a list of pods and checks that limits and requests are defined for both cpu and memory if I understand correctly

Comment (prudhvi-165) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:330]: Updated

File: tests/Tests/100__deploy/100__installation/102__post_install.robot (hunk 2 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot]: The separate keyword is removed and changes are combined into `Verify CPU And Memory Requests And Limits Are Defined For All Containers In All Pods in Project`

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/100__deploy/100__installation/102__post_install.robot [hunk 4]
```diff
@@ -296,6 +297,12 @@ Verify CPU And Memory Requests And Limits Are Defined For All Containers In All
     ${project_pods_info}=    Fetch Project Pods Info    ${project}
     FOR    ${pod_info}    IN    @{project_pods_info}
         Verify CPU And Memory Requests And Limits Are Defined For Pod    ${pod_info}
```
[the remark this answers is not in the document]
Comment (pablofelix) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:299]: I really liked the idea of adding  `expected-cpu-request` and `expected-memory-request` as optional parameters  for `Verify CPU And Memory Requests And Limits Are Defined For Pod`  keyword but while I was moving forward with the implementation, the keyword  was gaining a lot of complexity and the code was becoming really ugly. I will work  then on your second suggestion and create a new one to check the expected values that can be reused. Once I finished Prudhvi will be able to use them in his test.

[the remark this answers is not in the document]
Comment (jgarciao) [reply tests/Tests/100__deploy/100__installation/102__post_install.robot:299]: Agreement: we'll merge this PR as it is and update the code once the new keyword is present

**Review by lugi0 — APPROVED** [review]

**Pull request merged**